### DISCUSSION TITLE
Uploading static vector through linker script

### DISCUSSION THREAD

**Discussion by Giorgio122** [stnolting/neorv32#1076] [Q&A]
Hello,
I have a question regarding the linker script. I have a series of bytes that have to be stored starting from a fixed memory address in the ddr of the device that I am using. However, for now i have no way of initializing it during boot phase (i am currently working to find a solution to this) so i was wondering if i could upload this data through the AXI interface of the neorv. However, receiving this data from uart and forwarding it toward the AXI interface requires to much time, and it is to big to be stored in the BRAMS of the device (i don't have enough).
I was just wondering, what happens if i modify the linker script by adding the vector at a fixed position in memory (a void one, such as 0xF0000000)? Will it be forwarded towards the AXI interface since it points to a void memory location and thus propagate towards the external resource to which the risc is connected? Is this a thing or I have to find another way?
Thanks!

**Comment by stnolting** [comment]
This is a very hard question as this highly depends on your actual use case...

> I was just wondering, what happens if i modify the linker script by adding the vector at a fixed position in memory (a void one, such as 0xF0000000)? Will it be forwarded towards the AXI interface since it points to a void memory location and thus propagate towards the external resource to which the risc is connected? Is this a thing or I have to find another way?

Hmm, good question...
This is how I would to do it (with minimal effort 😅).

Declare your data vector (could also be an array) as a global constant somewhere in your C sources:

```c
const uint32_t your_vector = 0x12345678;
```

Then you could define a constructor (getting executed before `main()`) that copies your vector to the desired target address:

```c
void __attribute__((destructor)) copy_your_vector(void) {

  uint32_t* dst = 0xABCD0000; // this is the destination address of your vector in DDR
  uint32_t* src = (uint32_t*)(&your_vector[0]);
  for (i=0; i<(sizeof()/4); i++) {
    *dst++ = *src++;
  }
}
```

This is not perfect, but it should do the job I think.

**Reply by Giorgio122** [reply]
Thanks for the reply... however, if I declare it in a static way then it will still be temporary loaded in the memory of the neorv right? But the problem is that the vector should contain more than 1MB of data. That is why I was trying to found if there was another way to feed the data without actually having it stored in the risc memory

**Reply by stnolting** [reply]
Oh ok, I understand.

Do you use openOCD/GDB for uploading the executable?
GDB can also initialize custom sections (at custom addresses).

**Reply by Giorgio122** [reply]
No, I am not using GBD. But maybe the best way, if i cannot initialize the ddr of the board that I am using from a file, is to find a workaround without using the neorv32. The question arise because I thought that being the neorv32 memory mapped, in some way I had the chance of directly communicating with the DDR while uploading the executable file.

**Reply by stnolting** [reply]
Of course the processor can communicate with the DDR memory. You just need some additional code for that.

Have a look at the XIP example program. There we are using a simple function to upload a data file via UART and to write it to the SPI flash.

https://github.com/stnolting/neorv32/blob/948a8e3f67184283150cbe2ff06f2c296ea5300b/sw/example/demo_xip/main.c#L146-L147

I think you could use a similar approach here.

GDB comes quite handy as it has full access to the entire address space - so it can write data to arbitrary locations without any software overhead.

**Reply by Giorgio122** [reply]
Thanks, i will have a look into that! Hopefully it would be the solution
